src/coreclr/jit/codegenlinear.cpp

```diff
@@ -1841,20 +1841,36 @@ void CodeGen::genPutArgStkFieldList(GenTreePutArgStk* putArgStk, unsigned outArg
 
     // Evaluate each of the GT_FIELD_LIST items into their register
     // and store their register into the outgoing argument area.
-    unsigned argOffset = putArgStk->getArgOffset();
+    const unsigned argOffset = putArgStk->getArgOffset();
     for (GenTreeFieldList::Use& use : putArgStk->gtOp1->AsFieldList()->Uses())
     {
         GenTree* nextArgNode = use.GetNode();
         genConsumeReg(nextArgNode);
 
-        regNumber reg  = nextArgNode->GetRegNum();
-        var_types type = nextArgNode->TypeGet();
-        emitAttr  attr = emitTypeSize(type);
+        regNumber reg             = nextArgNode->GetRegNum();
+        var_types type            = use.GetType();
+        unsigned  thisFieldOffset = argOffset + use.GetOffset();
 
-        // Emit store instructions to store the registers produced by the GT_FIELD_LIST into the outgoing
-        // argument area.
-        unsigned thisFieldOffset = argOffset + use.GetOffset();
-        GetEmitter()->emitIns_S_R(ins_Store(type), attr, reg, outArgVarNum, thisFieldOffset);
+// Emit store instructions to store the registers produced by the GT_FIELD_LIST into the outgoing
+// argument area.
+
+#ifdef FEATURE_SIMD
+#if defined(OSX_ARM64_ABI)
+        // storing of TYP_SIMD12 (i.e. Vector3) argument.
+        if (type == TYP_SIMD12)
+        {
+            // Need an additional integer register to extract upper 4 bytes from data.
+            regNumber tmpReg = nextArgNode->GetSingleTempReg();
+
+            GetEmitter()->emitStoreSIMD12ToLclOffset(outArgVarNum, thisFieldOffset, reg, tmpReg);
+        }
+        else
+#endif // OSX_ARM64_ABI
+#endif // FEATURE_SIMD
+        {
+            emitAttr attr = emitTypeSize(type);
+            GetEmitter()->emitIns_S_R(ins_Store(type), attr, reg, outArgVarNum, thisFieldOffset);
```

Comment:
@sandreenko Correct me if I am wrong - on non macos-arm64 `TYP_SIMD12` is going to take 16 bytes on the stack and the JIT uses `str Qd, [Xn]` instruction

Reply:
@echesakovMSFT there are different cases:
1. pass in regs;
2. pass on stack;
x
1. LCL_VAR
2. LCL_FLD
3. OBJ/IND
4. FIELD_LIST

usually, if we pass `LCL_VAR SIMD12`  we will use `str Qd, [Xn]`, because we know that LCL_VAR has 4 bytes after 12 bytes of SIMD that we can read, but if SIMD12 is a field of a struct then we will have to read exactly 12 bytes, so we will have `IND SIMD12(ADDR(LCL_FLD))` and it has similar logic.